example/common/flat_write_stream.hpp

```diff
@@ -0,0 +1,554 @@
+                static_assert(is_completion_handler<WriteHandler,
+                                      void(error_code, std::size_t)>::value,
+                              "WriteHandler requirements not met");
+
+                auto size = boost::asio::buffer_size(buffers);
+
+                std::cout << "write_some buffer size: " << size << std::endl;   //todo: remove
+
+                boost::beast::async_completion<WriteHandler,
+                        void(boost::beast::error_code, size_t)>
+                        init{handler};
+
+                if(size > 0) {
+                    auto iter = find_single(buffers);
+                    if(iter != buffers.end()) {
+                        write_some_op<
+                                ConstBufferSequence,
+                                boost::beast::handler_type<
+                                        WriteHandler, void(boost::beast::error_code, std::size_t)>> op(handler, *this, boost::asio::const_buffers_1{*iter}, std::shared_ptr<char>(nullptr));
+
+                        op(boost::beast::error_code{}, 0);
+                        return init.result.get();
+                    }
+                    else {
+                        if(size <= max_stack_size) {
+                            write_some_op<
+                                    ConstBufferSequence,
+                                    boost::beast::handler_type<
+                                            WriteHandler, void(boost::beast::error_code, std::size_t)>> op(handler, *this, stack_flatten(buffers), std::shared_ptr<char>(nullptr));
+
+                            op(boost::beast::error_code{}, 0);
+                            return init.result.get();
+                        } else {
+                            try {
+                                auto buf = std::shared_ptr<char>(new char[size], std::default_delete<char[]>());
+                                auto copy_size = boost::asio::buffer_copy(boost::asio::buffer(buf.get(), size),
+                                                                          buffers);
+                                write_some_op<
+                                        ConstBufferSequence,
+                                        boost::beast::handler_type<
+                                                WriteHandler, void(boost::beast::error_code, std::size_t)>> op(handler, *this, boost::asio::const_buffers_1(buf.get(), copy_size), buf);
+
+                                op(boost::beast::error_code{}, 0);
+                                return init.result.get();
+                            }
+                            catch (std::bad_alloc const &) {
+                                // todo
+                                return;
+                            }
+                        }
+                    }
+                }
+                else {
+                    return next_layer_.async_write_some(buffers, handler);
```

Comment:
If you're going to forward the call then you should not also construct `init`, or else you could be left with a moved-from handler.